### PR TITLE
[improvement](nereids) Composite predicate supports range predicate when rewritting by materialzied view

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes
It supports predicate composite as following:
materialized view define
>        select l_shipdate, o_orderdate, l_partkey, l_suppkey
>        from lineitem_1
>        left join orders_1
>        on lineitem_1.l_orderkey = orders_1.o_orderkey
>        where l_shipdate > '2023-10-19'

the query as following can be rewritten by the materialized view above
>        select l_shipdate, o_orderdate, l_partkey, l_suppkey
>        from lineitem_1
>        left join orders_1
>        on lineitem_1.l_orderkey = orders_1.o_orderkey
>        where l_shipdate > '2023-10-25'

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morrySnow** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

File: regression-test/suites/nereids_rules_p0/mv/dimension_equal/filter_equal_or_notequal.groovy [hunk 1]
```diff
@@ -158,132 +158,132 @@ suite("filter_equal_or_notequal_case") {
...
-        where l_shipdate != '2023-10-17'
-        """
-    explain {
-        sql("${query_sql}")
-        contains "${mv_name}(${mv_name})"
-    }
-    compare_res(query_sql + " order by 1,2,3,4")
-
-
+//    create_mv_lineitem(mv_name, mtmv_sql)
```
Comment (morrySnow) [inline regression-test/suites/nereids_rules_p0/mv/dimension_equal/filter_equal_or_notequal.groovy]: why comment these cases?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/SimplifyRange.java [hunk 2]
```diff
@@ -65,12 +65,12 @@
...
- *    for example:
- *    a > 1 and a > 2
- *    1. a > 1 => RangeValueDesc((1...+∞)), a > 2 => RangeValueDesc((2...+∞))
- *    2. (1...+∞) intersect (2...+∞) => (2...+∞)
+ * 1. extract conjunctions then build `ValueDesc` for each conjunction
+ * 2. grouping according to `reference`, `ValueDesc` in the same group can perform intersect
+ * for example:
+ * a > 1 and a > 2
+ * 1. a > 1 => RangeValueDesc((1...+∞)), a > 2 => RangeValueDesc((2...+∞))
+ * 2. (1...+∞) intersect (2...+∞) => (2...+∞)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/SimplifyRange.java:73]: add back Indentation

**Review by starocean999 — DISMISSED** [review]

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/SimplifyRangeTest.java [hunk 3]
```diff
@@ -86,7 +86,8 @@ public void testSimplify() {
         assertRewrite("(TA > 10 or TA > 20) and (TB > 10 and TB > 20)", "TA > 10 and TB > 20");
         assertRewrite("((TB > 30 and TA > 40) and TA > 20) and (TB > 10 and TB > 20)", "TB > 30 and TA > 40");
         assertRewrite("(TA > 10 and TB > 10) or (TB > 10 and TB > 20)", "TA > 10 and TB > 10 or TB > 20");
-        assertRewrite("((TA > 10 or TA > 5) and TB > 10) or (TB > 10 and (TB > 20 or TB < 10))", "(TA > 5 and TB > 10) or (TB > 10 and (TB > 20 or TB < 10))");
+        assertRewrite("((TA > 10 or TA > 5) and TB > 10) or (TB > 10 and (TB > 20 or TB < 10))",
+                "(TA > 5 and TB > 10) or (TB > 10 and (TB > 20 or TB < 10))");
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/SimplifyRangeTest.java:90]: revert unnecessary format

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/SimplifyRangeTest.java [hunk 4]
```diff
@@ -158,6 +164,72 @@ public void testSimplify() {
...
+                "TA in (date '2024-01-01',date '2024-01-02',date '2024-01-03') or TA < date '2024-01-01'");
+        assertRewrite("TA in (date '2024-01-01',date '2024-01-02') or TA in (date '2024-01-02', date '2024-01-03')",
+                "TA in (date '2024-01-01',date '2024-01-02',date '2024-01-03')");
+        assertRewrite("TA in (date '2024-01-01',date '2024-01-02') and TA in (date '2024-01-03', date '2024-01-04')",
+                "FALSE");
+        assertRewrite("TA = date '2024-01-03' and TA = date '2024-01-01'", "FALSE");
+        assertRewrite("TA in (date '2024-01-01') and TA in (date '2024-01-03')", "FALSE");
+        assertRewrite("TA in (date '2024-01-03') and TA in (date '2024-01-03')", "TA = date '2024-01-03'");
+        assertRewrite("(TA > date '2024-01-03' and TA < date '2024-01-01') and TB < date '2024-01-05'", "FALSE");
+        assertRewrite("(TA > date '2024-01-03' and TA < date '2024-01-01') or TB < date '2024-01-05'",
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/SimplifyRangeTest.java:229]: add datetime test case

**Comment by seawinde** [comment]
run buildall

File: regression-test/suites/nereids_rules_p0/mv/dimension_equal/filter_equal_or_notequal.groovy (hunk 1 above)
Comment (seawinde) [reply regression-test/suites/nereids_rules_p0/mv/dimension_equal/filter_equal_or_notequal.groovy]: should add, will fixed

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/SimplifyRangeTest.java (hunk 3 above)
Comment (seawinde) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/SimplifyRangeTest.java:90]: ok

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by keanji-x — APPROVED** [review]

**Pull request merged**